src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -336,7 +336,16 @@ internal static bool IsCompilerServerSupported(string tempPath)
                 Log("Attempt to connect named pipe '{0}'", pipeName);
                 try
                 {
-                    await pipeStream.ConnectAsync(timeoutMs, cancellationToken).ConfigureAwait(false);
+                    // NamedPipeClientStream.ConnectAsync on the "full" framework has a bug where it
+                    // tries to move potentially expensive work (actually connecting to the pipe) to
+                    // a background thread with Task.Factory.StartNew. However, that call will merely
+                    // queue the work onto the TaskScheduler associated with the "current" Task which
+                    // does not guarantee it will be processed on a background thread and this could
+                    // lead to a hang.
+                    // To avoid this, we first force ourselves to a background thread using Task.Run.
+                    // This ensures that the Task created by ConnectAsync will run on the default
+                    // TaskScheduler (i.e., on a threadpool thread) which was the intent all along.
+                    await Task.Run(async () => await pipeStream.ConnectAsync(timeoutMs, cancellationToken).ConfigureAwait(false)).ConfigureAwait(false);
```

Comment:
@stephentoub FYI

Reply:
The async/await shouldn't be necessary; you can do:
```C#
await Task.Run(() => pipeStream.ConnectAsync(timeoutMS, CancellationToken)).ConfigureAwait(false);
```

Reply:
I always get confused on this point. The `Task` created by `Task.Run` will not complete until the `Task` returned by `ConnectAsync`, correct?

Reply:
Correct.